Add Python style guide (#5)
C0D1UM/technical-standard: django/python-style-guide/README.md

# Python Style Guide

## Base Style guide

We follow [Google style guide for Python](https://google.github.io/styleguide/pyguide.html). The rest of the document describes additions and clarifications.

## Linting

> * You can use `pylint` with this [pylintc](https://google.github.io/styleguide/pylintrc) to check if your code is passed standard.

Reviewer: if your code passed the standard.

## Formatter

* You can use [yapf](https://github.com/google/yapf/) to avoid arguing over formatting.

> * For Pycharm user you can configure `yapf` as external tools see [How to config Pycharm](PYCHARM.md).

Reviewer: For PyCharm users, you can configure `yapf` as an external tool, see [How to config PyCharm](PYCHARM.md)

> * For VSCode user in `setting.json` set `python.formatting.provider: "yapf"`

Reviewer: For VSCode users, in `setting.json` set `python.formatting.provider: "yapf"`

## Imports

Imports should be grouped in the following order

> * stdlib

Reviewer 2: Maybe krub?
```suggestion
* standard libraries
```

* third party libraries
* local modules

> Each group should be sorted lexicographically, ignoring case,separate by a newline, e.g.

Reviewer 2: What do you think if we change to this krubb?

```suggestion
Each group should be sorted lexicographically, ignoring case, separate by an empty line, e.g.
```

```python
# Yes
if seq:
if not seq:

# No
if len(seq):
if len(seq) == 0:
```

## Trailing commas

Trailing commas in sequences of items are recommended only when the closing container token ], ), or } does not appear on the same line as the final element. The presence of a trailing comma is also used as a hint to our Python code auto-formatter YAPF to direct it to auto-format the container of items to one item per line when the , after the final element is present.

```python
# Yes
golomb3 = [0, 1, 3]
golomb4 = [
    0,
    1,
    4,
    6,
]

# No
golomb4 = [
    0,
    1,
    4,
    6
]
```

## Joining Path

```python
import os

# Yes
data_path = os.path.join(data_folder, "data.csv")

# No
data_path = data_folder + "/data.csv"
```

## Logging

Prefer logging over print statement

```python
import logging

# Yes
logger = logging.getLogger(__name__)

if bad_mojo:
    logger.error('Something went wrong!')

# No
if bad_mojo:
    print('Something went wrong!')
```

## With statement

Always open file with context manager

```python
# Yes
with open('file.txt') as file:
    pass

# No
file = open('file.txt')
file.close()
```

> For psycopg2 connection you should use [try catch](https://www.psycopg.org/docs/usage.html#with-statement) style

Reviewer: Punctuation - Comma after connection.
For psycopg2 connection, you should use .. style.

```python
conn = psycopg2.connect(DSN)

> # Yes

Reviewer: Could it combine like this na krub:
```python
conn = psycopg2.connect(DSN)

try:
    with conn.cursor() as curs:
        curs.execute(SQL2)
finally:
     conn.close()
```

Because during SCB I read somewhere that it's better to use context manager so that it handle many things by itself, instead of handle by our own. I couldn't find it anymore.

Reviewer: I think I remember it wrong, after talk with Ten, it's when you do transaction atomic, so this one resolve krub.
